File: src/main/java/com/owncloud/android/media/MediaControlView.java
/**
 * ownCloud Android client application
 *
 * @author David A. Velasco
 * Copyright (C) 2015 ownCloud Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.owncloud.android.media;

import android.content.Context;
import android.media.MediaPlayer;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.MediaController.MediaPlayerControl;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

import com.owncloud.android.R;
import com.owncloud.android.utils.ThemeUtils;

import java.util.Formatter;
import java.util.Locale;


/**
 * View containing controls for a {@link MediaPlayer}. 
 *
 * Holds buttons "play / pause", "rewind", "fast forward" 
 * and a progress slider. 
 *
 * It synchronizes itself with the state of the 
 * {@link MediaPlayer}.
 */

public class MediaControlView extends FrameLayout /* implements OnLayoutChangeListener, OnTouchListener */ implements OnClickListener, OnSeekBarChangeListener {

    private MediaPlayerControl mPlayer;
    private Context mContext;
    private View mRoot;
    private ProgressBar mProgress;
    private TextView mEndTime, mCurrentTime;
    private boolean mDragging;
    private static final int SHOW_PROGRESS = 1;
    private ImageButton mPauseButton;
    private ImageButton mFfwdButton;
    private ImageButton mRewButton;

    public MediaControlView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;

        FrameLayout.LayoutParams frameParams = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
        );
        LayoutInflater inflate = (LayoutInflater) mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        mRoot = inflate.inflate(R.layout.media_control, null);
        initControllerView(mRoot);
        addView(mRoot, frameParams);

        setFocusable(true);
        setFocusableInTouchMode(true);
        setDescendantFocusability(ViewGroup.FOCUS_AFTER_DESCENDANTS);
        requestFocus();
    }

    @Override
    public void onFinishInflate() {
        super.onFinishInflate();
        /*
        if (mRoot != null)
            initControllerView(mRoot);
         */
    }

    /* TODO REMOVE
    public MediaControlView(Context context, boolean useFastForward) {
        super(context);
        mContext = context;
        mUseFastForward = useFastForward;
        initFloatingWindowLayout();
        //initFloatingWindow();
    }
    */

    /* TODO REMOVE
    public MediaControlView(Context context) {
        this(context, true);
    }
    */
    
    /* T
    private void initFloatingWindow() {
        mWindowManager = (WindowManager)mContext.getSystemService(Context.WINDOW_SERVICE);
        mWindow = PolicyManager.makeNewWindow(mContext);
        mWindow.setWindowManager(mWindowManager, null, null);
        mWindow.requestFeature(Window.FEATURE_NO_TITLE);
        mDecor = mWindow.getDecorView();
        mDecor.setOnTouchListener(mTouchListener);
        mWindow.setContentView(this);
        mWindow.setBackgroundDrawableResource(android.R.color.transparent);
        
        // While the media controller is up, the volume control keys should
        // affect the media stream type
        mWindow.setVolumeControlStream(AudioManager.STREAM_MUSIC);

        setFocusable(true);
        setFocusableInTouchMode(true);
        setDescendantFocusability(ViewGroup.FOCUS_AFTER_DESCENDANTS);
        requestFocus();
    }
    */

    /*
    // Allocate and initialize the static parts of mDecorLayoutParams. Must
    // also call updateFloatingWindowLayout() to fill in the dynamic parts
    // (y and width) before mDecorLayoutParams can be used.
    private void initFloatingWindowLayout() {
        mDecorLayoutParams = new WindowManager.LayoutParams();
        WindowManager.LayoutParams p = mDecorLayoutParams;
        p.gravity = Gravity.TOP;
        p.height = LayoutParams.WRAP_CONTENT;
        p.x = 0;
        p.format = PixelFormat.TRANSLUCENT;
        p.type = WindowManager.LayoutParams.TYPE_APPLICATION_PANEL;
        p.flags |= WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM
                | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                | WindowManager.LayoutParams.FLAG_SPLIT_TOUCH;
        p.token = null;
        p.windowAnimations = 0; // android.R.style.DropDownAnimationDown;
    }
    */

    // Update the dynamic parts of mDecorLayoutParams
    // Must be called with mAnchor != NULL.
    /*
    private void updateFloatingWindowLayout() {
        int [] anchorPos = new int[2];
        mAnchor.getLocationOnScreen(anchorPos);

        WindowManager.LayoutParams p = mDecorLayoutParams;
        p.width = mAnchor.getWidth();
        p.y = anchorPos[1] + mAnchor.getHeight();
    }
    */

    /*
    // This is called whenever mAnchor's layout bound changes
    public void onLayoutChange(View v, int left, int top, int right,
            int bottom, int oldLeft, int oldTop, int oldRight,
            int oldBottom) {
        //updateFloatingWindowLayout();
        if (mShowing) {
            mWindowManager.updateViewLayout(mDecor, mDecorLayoutParams);
        }
    }
    */
    
    /*
    public boolean onTouch(View v, MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            if (mShowing) {
                hide();
            }
        }
            return false;
    }
    */


    public void setMediaPlayer(MediaPlayerControl player) {
        mPlayer = player;
        mHandler.sendEmptyMessage(SHOW_PROGRESS);
        updatePausePlay();
    }

    public void stopMediaPlayerMessages() {
        mHandler.removeMessages(SHOW_PROGRESS);
    }


    private void initControllerView(View v) {
        mPauseButton = (ImageButton) v.findViewById(R.id.playBtn);
        if (mPauseButton != null) {
            mPauseButton.requestFocus();
            mPauseButton.setOnClickListener(this);
        }

        mFfwdButton = (ImageButton) v.findViewById(R.id.forwardBtn);
        if (mFfwdButton != null) {
            mFfwdButton.setOnClickListener(this);
        }

        mRewButton = (ImageButton) v.findViewById(R.id.rewindBtn);
        if (mRewButton != null) {
            mRewButton.setOnClickListener(this);
        }

        mProgress = (ProgressBar) v.findViewById(R.id.progressBar);
        if (mProgress != null) {
            if (mProgress instanceof SeekBar) {
                SeekBar seeker = (SeekBar) mProgress;
                ThemeUtils.colorHorizontalSeekBar(seeker);
                seeker.setOnSeekBarChangeListener(this);
            } else {
                ThemeUtils.colorHorizontalProgressBar(mProgress, ThemeUtils.primaryAccentColor());
            }
            mProgress.setMax(1000);
        }

        mEndTime = (TextView) v.findViewById(R.id.totalTimeText);
        mCurrentTime = (TextView) v.findViewById(R.id.currentTimeText);
    }


    /**
     * Disable pause or seek buttons if the stream cannot be paused or seeked.
     * This requires the control interface to be a MediaPlayerControlExt
     */
    private void disableUnsupportedButtons() {
        try {
            if (mPauseButton != null && !mPlayer.canPause()) {
                mPauseButton.setEnabled(false);
            }
            if (mRewButton != null && !mPlayer.canSeekBackward()) {
                mRewButton.setEnabled(false);
            }
            if (mFfwdButton != null && !mPlayer.canSeekForward()) {
                mFfwdButton.setEnabled(false);
            }
        } catch (IncompatibleClassChangeError ex) {
            // We were given an old version of the interface, that doesn't have
            // the canPause/canSeekXYZ methods. This is OK, it just means we
            // assume the media can be paused and seeked, and so we don't disable
            // the buttons.
        }
    }


    private Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            int pos;
            if (msg.what == SHOW_PROGRESS) {
                pos = setProgress();
                if (!mDragging) {
                    msg = obtainMessage(SHOW_PROGRESS);
                    sendMessageDelayed(msg, 1000 - (pos % 1000));
                }
            }
        }
    };

    private String stringForTime(int timeMs) {
        int totalSeconds = timeMs / 1000;

        int seconds = totalSeconds % 60;
        int minutes = (totalSeconds / 60) % 60;
        int hours = totalSeconds / 3600;

        final StringBuilder mFormatBuilder = new StringBuilder();
        final Formatter mFormatter = new Formatter(mFormatBuilder, Locale.getDefault());
        if (hours > 0) {
            return mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString();
        } else {
            return mFormatter.format("%02d:%02d", minutes, seconds).toString();
        }
    }

    private int setProgress() {
        if (mPlayer == null || mDragging) {
            return 0;
        }
        int position = mPlayer.getCurrentPosition();
        int duration = mPlayer.getDuration();
        if (mProgress != null) {
            if (duration > 0) {
                // use long to avoid overflow
                long pos = 1000L * position / duration;
                mProgress.setProgress((int) pos);
            }
            int percent = mPlayer.getBufferPercentage();
            mProgress.setSecondaryProgress(percent * 10);
        }

        if (mEndTime != null) {
            mEndTime.setText(stringForTime(duration));
        }
        if (mCurrentTime != null) {
            mCurrentTime.setText(stringForTime(position));
        }
        return position;
    }


    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        int keyCode = event.getKeyCode();
        final boolean uniqueDown = event.getRepeatCount() == 0
                && event.getAction() == KeyEvent.ACTION_DOWN;
        if (keyCode == KeyEvent.KEYCODE_HEADSETHOOK
                || keyCode == KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE
                || keyCode == KeyEvent.KEYCODE_SPACE) {
            if (uniqueDown) {
                doPauseResume();
                //show(sDefaultTimeout);
                if (mPauseButton != null) {
                    mPauseButton.requestFocus();
                }
            }
            return true;
        } else if (keyCode == KeyEvent.KEYCODE_MEDIA_PLAY) {
            if (uniqueDown && !mPlayer.isPlaying()) {
                mPlayer.start();
                updatePausePlay();
                //show(sDefaultTimeout);
            }
            return true;
        } else if (keyCode == KeyEvent.KEYCODE_MEDIA_STOP
                || keyCode == KeyEvent.KEYCODE_MEDIA_PAUSE) {
            if (uniqueDown && mPlayer.isPlaying()) {
                mPlayer.pause();
                updatePausePlay();
                //show(sDefaultTimeout);
            }
            return true;
        }

        //show(sDefaultTimeout);
        return super.dispatchKeyEvent(event);
    }

    public void updatePausePlay() {
        if (mRoot == null || mPauseButton == null) {
            return;
        }

        if (mPlayer.isPlaying()) {
            mPauseButton.setImageResource(android.R.drawable.ic_media_pause);
        } else {
            mPauseButton.setImageResource(android.R.drawable.ic_media_play);
        }
    }

    private void doPauseResume() {
        if (mPlayer.isPlaying()) {
            mPlayer.pause();
        } else {
            mPlayer.start();
        }
        updatePausePlay();
    }

    @Override
    public void setEnabled(boolean enabled) {
        if (mPauseButton != null) {
            mPauseButton.setEnabled(enabled);
        }
        if (mFfwdButton != null) {
            mFfwdButton.setEnabled(enabled);
        }
        if (mRewButton != null) {
            mRewButton.setEnabled(enabled);
        }
        if (mProgress != null) {
            mProgress.setEnabled(enabled);
        }
        disableUnsupportedButtons();
        super.setEnabled(enabled);
    }

    @Override
    public void onClick(View v) {
        int pos;
        boolean playing = mPlayer.isPlaying();
        switch (v.getId()) {

            case R.id.playBtn:
                doPauseResume();
                break;

            case R.id.rewindBtn:
                pos = mPlayer.getCurrentPosition();
                pos -= 5000;
                mPlayer.seekTo(pos);
                if (!playing) {
                    mPlayer.pause();  // necessary in some 2.3.x devices
                }
                setProgress();
                break;

            case R.id.forwardBtn:
                pos = mPlayer.getCurrentPosition();
                pos += 15000;
                mPlayer.seekTo(pos);
                if (!playing) {
                    mPlayer.pause(); // necessary in some 2.3.x devices
                }
                setProgress();
                break;
        }
    }


    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (!fromUser) {
            // We're not interested in programmatically generated changes to
            // the progress bar's position.
            return;
        }

        long duration = mPlayer.getDuration();
        long newposition = (duration * progress) / 1000L;
        mPlayer.seekTo((int) newposition);
        if (mCurrentTime != null) {
            mCurrentTime.setText(stringForTime((int) newposition));
        }
    }

    /**
     * Called in devices with touchpad when the user starts to adjust the 
     * position of the seekbar's thumb.
     *
     * Will be followed by several onProgressChanged notifications.
     */
    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        mDragging = true;                           // monitors the duration of dragging 
        mHandler.removeMessages(SHOW_PROGRESS);     // grants no more updates with media player progress while dragging 
    }


    /**
     * Called in devices with touchpad when the user finishes the
     * adjusting of the seekbar.
     */
    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        mDragging = false;
        setProgress();
        updatePausePlay();
        mHandler.sendEmptyMessage(SHOW_PROGRESS);    // grants future updates with media player progress 
    }

    @Override
    public void onInitializeAccessibilityEvent(AccessibilityEvent event) {
        super.onInitializeAccessibilityEvent(event);
        event.setClassName(MediaControlView.class.getName());
    }

    @Override
    public void onInitializeAccessibilityNodeInfo(AccessibilityNodeInfo info) {
        super.onInitializeAccessibilityNodeInfo(info);
        info.setClassName(MediaControlView.class.getName());
    }

}

File: app/src/main/java/com/owncloud/android/media/MediaControlView.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2022 Álvaro Brey Vilas <alvaro@alvarobrey.com>
 * SPDX-FileCopyrightText: 2018-2020 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2019 Chris Narkiewicz <hello@ezaquarii.com>
 * SPDX-FileCopyrightText: 2018 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2015 ownCloud Inc.
 * SPDX-FileCopyrightText: 2013 David A. Velasco <dvelasco@solidgear.es>
 * SPDX-License-Identifier: GPL-2.0-only AND AGPL-3.0-or-later
 */
package com.owncloud.android.media

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.core.content.ContextCompat
import androidx.media3.common.Player
import com.owncloud.android.MainApp
import com.owncloud.android.R
import com.owncloud.android.databinding.MediaControlBinding
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.utils.theme.ViewThemeUtils
import java.util.Formatter
import java.util.Locale
import javax.inject.Inject

/**
 * View containing controls for a MediaPlayer.
 *
 *
 * Holds buttons "play / pause", "rewind", "fast forward" and a progress slider.
 *
 *
 * It synchronizes itself with the state of the MediaPlayer.
 */
class MediaControlView(context: Context, attrs: AttributeSet?) :
    LinearLayout(context, attrs),
    View.OnClickListener,
    OnSeekBarChangeListener {

    private var playerControl: Player? = null
    private var binding: MediaControlBinding
    private var isDragging = false

    @Inject
    lateinit var viewThemeUtils: ViewThemeUtils

    public override fun onFinishInflate() {
        super.onFinishInflate()
    }

    @Suppress("MagicNumber")
    fun setMediaPlayer(player: Player?) {
        playerControl = player
        handler.sendEmptyMessage(SHOW_PROGRESS)

        handler.postDelayed({
            updatePausePlay()
            setProgress()
        }, 100)
    }

    fun stopMediaPlayerMessages() {
        handler.removeMessages(SHOW_PROGRESS)
    }

    @Suppress("MagicNumber")
    private fun initControllerView() {
        binding.playBtn.requestFocus()

        binding.playBtn.setOnClickListener(this)
        binding.forwardBtn.setOnClickListener(this)
        binding.rewindBtn.setOnClickListener(this)

        binding.progressBar.run {
            viewThemeUtils.platform.themeHorizontalSeekBar(this)
            setMax(1000)
        }

        binding.progressBar.setOnSeekBarChangeListener(this)

        viewThemeUtils.material.run {
            colorMaterialButtonPrimaryTonal(binding.rewindBtn)
            colorMaterialButtonPrimaryTonal(binding.playBtn)
            colorMaterialButtonPrimaryTonal(binding.forwardBtn)
        }
    }

    /**
     * Disable pause or seek buttons if the stream cannot be paused or seeked.
     * This requires the control interface to be a MediaPlayerControlExt
     */
    private fun disableUnsupportedButtons() {
        try {
            if (playerControl!!.isCommandAvailable(Player.COMMAND_PLAY_PAUSE).not()) {
                binding.playBtn.isEnabled = false
            }

            if (playerControl!!.isCommandAvailable(Player.COMMAND_SEEK_BACK).not()) {
                binding.rewindBtn.isEnabled = false
            }
            if (playerControl!!.isCommandAvailable(Player.COMMAND_SEEK_FORWARD).not()) {
                binding.forwardBtn.isEnabled = false
            }
        } catch (ex: IncompatibleClassChangeError) {
            // We were given an old version of the interface, that doesn't have
            // the canPause/canSeekXYZ methods. This is OK, it just means we
            // assume the media can be paused and seeked, and so we don't disable
            // the buttons.
            Log_OC.i(TAG, "Old media interface detected")
        }
    }

    @Suppress("MagicNumber")
    private val handler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            if (msg.what == SHOW_PROGRESS) {
                updatePausePlay()
                val pos = setProgress()

                if (!isDragging) {
                    sendMessageDelayed(obtainMessage(SHOW_PROGRESS), (1000 - pos % 1000).toLong())
                }
            }
        }
    }

    init {
        MainApp.getAppComponent().inject(this)

        val inflate = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        binding = MediaControlBinding.inflate(inflate, this, true)
        initControllerView()
        isFocusable = true
        setFocusableInTouchMode(true)
        setDescendantFocusability(FOCUS_AFTER_DESCENDANTS)
        requestFocus()
    }

    @Suppress("MagicNumber")
    private fun formatTime(timeMs: Long): String {
        val totalSeconds = timeMs / 1000
        val seconds = totalSeconds % 60
        val minutes = totalSeconds / 60 % 60
        val hours = totalSeconds / 3600
        val mFormatBuilder = StringBuilder()
        val mFormatter = Formatter(mFormatBuilder, Locale.getDefault())
        return if (hours > 0) {
            mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString()
        } else {
            mFormatter.format("%02d:%02d", minutes, seconds).toString()
        }
    }

    @Suppress("MagicNumber")
    private fun setProgress(): Long {
        var position = 0L
        if (playerControl == null || isDragging) {
            position = 0
        }

        playerControl?.let { playerControl ->
            position = playerControl.currentPosition
            val duration = playerControl.duration
            if (duration > 0) {
                // use long to avoid overflow
                val pos = 1000L * position / duration
                binding.progressBar.progress = pos.toInt()
            }
            val percent = playerControl.bufferedPercentage
            binding.progressBar.setSecondaryProgress(percent * 10)
            val endTime = if (duration > 0) formatTime(duration) else "--:--"
            binding.totalTimeText.text = endTime
            binding.currentTimeText.text = formatTime(position)
        }

        return position
    }

    @Suppress("ReturnCount")
    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        val keyCode = event.keyCode
        val uniqueDown = (event.repeatCount == 0 && event.action == KeyEvent.ACTION_DOWN)

        when (keyCode) {
            KeyEvent.KEYCODE_HEADSETHOOK, KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE, KeyEvent.KEYCODE_SPACE -> {
                if (uniqueDown) {
                    doPauseResume()
                    // show(sDefaultTimeout);
                    binding.playBtn.requestFocus()
                }
                return true
            }

            KeyEvent.KEYCODE_MEDIA_PLAY -> {
                if (uniqueDown && playerControl?.playWhenReady == false) {
                    playerControl?.play()
                    updatePausePlay()
                }
                return true
            }

            KeyEvent.KEYCODE_MEDIA_STOP,
            KeyEvent.KEYCODE_MEDIA_PAUSE
            -> {
                if (uniqueDown && playerControl?.playWhenReady == true) {
                    playerControl?.pause()
                    updatePausePlay()
                }
                return true
            }

            else -> return super.dispatchKeyEvent(event)
        }
    }

    fun updatePausePlay() {
        binding.playBtn.icon = ContextCompat.getDrawable(
            context,
            // isPlaying reflects if the playback is actually moving forward, If the media is buffering and it will play when ready
            // it would still return that it is not playing. So, in case of buffering it will show the pause icon which would show that
            // media is loading, when user has not paused but moved the progress to a different position this works as a buffering signal.
            if (playerControl?.isPlaying == true) {
                R.drawable.ic_pause
            } else {
                R.drawable.ic_play
            }
        )
        binding.forwardBtn.visibility = if (playerControl!!.isCommandAvailable(Player.COMMAND_SEEK_FORWARD)) {
            VISIBLE
        } else {
            INVISIBLE
        }
        binding.rewindBtn.visibility = if (playerControl!!.isCommandAvailable(Player.COMMAND_SEEK_BACK)) {
            VISIBLE
        } else {
            INVISIBLE
        }
    }

    private fun doPauseResume() {
        playerControl?.run {
            if (playWhenReady) {
                pause()
            } else {
                play()
            }
        }
        updatePausePlay()
    }

    override fun setEnabled(enabled: Boolean) {
        binding.playBtn.setEnabled(enabled)
        binding.forwardBtn.setEnabled(enabled)
        binding.rewindBtn.setEnabled(enabled)
        binding.progressBar.setEnabled(enabled)

        disableUnsupportedButtons()

        super.setEnabled(enabled)
    }

    @Suppress("MagicNumber")
    override fun onClick(v: View) {

        playerControl?.let { playerControl ->
            val playing = playerControl.playWhenReady
            val id = v.id

            when (id) {
                R.id.playBtn -> {
                    doPauseResume()
                }

                R.id.rewindBtn -> {
                    playerControl.seekBack()
                    if (!playing) {
                        playerControl.pause() // necessary in some 2.3.x devices
                    }
                    setProgress()
                }

                R.id.forwardBtn -> {
                    playerControl.seekForward()
                    if (!playing) {
                        playerControl.pause() // necessary in some 2.3.x devices
                    }

                    setProgress()
                }

                else -> {
                }
            }
        }
    }

    @Suppress("MagicNumber")
    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        if (!fromUser) {
            // We're not interested in programmatically generated changes to
            // the progress bar's position.
            return
        }

        playerControl?.let { playerControl ->
            val duration = playerControl.duration.toLong()
            val newPosition = duration * progress / 1000L
            playerControl.seekTo(newPosition)
            binding.currentTimeText.text = formatTime(newPosition)
        }
    }

    /**
     * Called in devices with touchpad when the user starts to adjust the position of the seekbar's thumb.
     *
     * Will be followed by several onProgressChanged notifications.
     */
    override fun onStartTrackingTouch(seekBar: SeekBar) {
        isDragging = true // monitors the duration of dragging
        handler.removeMessages(SHOW_PROGRESS) // grants no more updates with media player progress while dragging
    }

    /**
     * Called in devices with touchpad when the user finishes the adjusting of the seekbar.
     */
    override fun onStopTrackingTouch(seekBar: SeekBar) {
        isDragging = false
        setProgress()
        updatePausePlay()
        handler.sendEmptyMessage(SHOW_PROGRESS) // grants future updates with media player progress
    }

    override fun onInitializeAccessibilityEvent(event: AccessibilityEvent) {
        super.onInitializeAccessibilityEvent(event)
        event.setClassName(MediaControlView::class.java.getName())
    }

    override fun onInitializeAccessibilityNodeInfo(info: AccessibilityNodeInfo) {
        super.onInitializeAccessibilityNodeInfo(info)
        info.setClassName(MediaControlView::class.java.getName())
    }

    companion object {
        private val TAG = MediaControlView::class.java.getSimpleName()
        private const val SHOW_PROGRESS = 1
    }
}
